File: rust/src/leetcode/Q84_largest_rectangle_in_histogram.rs
/*
 * @lc app=leetcode id=84 lang=rust
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut result: i32 = 0;
        let mut stack: Vec<(usize, i32)> = vec![];

        for (i, h) in heights.iter().enumerate() {
            let mut start: usize = i;

            while let Some(&(index, height)) = stack.last() {
                if h <= &height {
                    stack.pop();

                    if height * (i - index) as i32 > result {
                        result = height * (i - index) as i32;
                    }

                    start = index;
                } else {
                    break
                }
            }

            stack.push((start, *h));
        }

        for (index, height) in stack {
            if height * (heights.len() - index) as i32 > result {
                result = height * (heights.len() - index) as i32;
            }
        }

        result
    }
}
// @lc code=end

File: python/Q84_largest_rectangle_in_histogram.py
#
# @lc app=leetcode id=84 lang=python3
#
# [84] Largest Rectangle in Histogram
#


# @lc code=start
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        max_area: int = 0
        stack: list[tuple[int, int]] = []  # (index, height)

        for i, h in enumerate(heights):
            start: int = i

            while stack and h < stack[-1][1]:
                index, height = stack.pop()
                max_area = max(max_area, height * (i - index))
                start = index

            stack.append((start, h))

        for i, h in stack:
            max_area = max(max_area, h * (len(heights) - i))

        return max_area


# @lc code=end
